### ISSUE TITLE
IIOException "Unsupported Image Type" while reading CMYK JPEG on OpenJDK

### ISSUE THREAD

**Issue by pszemus** [haraldk/TwelveMonkeys#341]
I've got `javax.imageio.IIOException: Unsupported Image Type` while reading JPG image with CMYK color space with `imageio-jpeg-3.3.2`.
Input image:
![1309390204204_0x22111122_cmyk](https://cloud.githubusercontent.com/assets/13220781/25179943/473c774a-250c-11e7-9e54-62d816fb6d58.jpeg)

**Comment by haraldk** [comment]
Hi pszemus,

Do you have a stack trace? 

I seem to be able to read this image just fine here. 

PS: The image seems to YCbCrA, not CMYK though. Blue bus, white text, green 1. 

-- 
Harald K

**Comment by pszemus** [comment]
Here's the stack trace:
```
javax.imageio.IIOException: Unsupported Image Type
        at com.sun.imageio.plugins.jpeg.JPEGImageReader.readInternal(JPEGImageReader.java:1068)
        at com.sun.imageio.plugins.jpeg.JPEGImageReader.read(JPEGImageReader.java:1039)
        at com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader.read(JPEGImageReader.java:382)
        at javax.imageio.ImageIO.read(ImageIO.java:1448)
        at javax.imageio.ImageIO.read(ImageIO.java:1400)
        at pl.wp.imageresizer.worker.client.integration.ResizeIT.createImage(ResizeIT.java:350)
        at pl.wp.imageresizer.worker.client.integration.ResizeIT.shouldResizeWithAutoCrop(ResizeIT.java:231)
        at pl.wp.imageresizer.worker.client.integration.ResizeIT.shouldResizeWithAutoCrop(ResizeIT.java:209)
        at pl.wp.imageresizer.worker.client.integration.ResizeIT.test(ResizeIT.java:55)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at junitparams.JUnitParamsRunner.runChild(JUnitParamsRunner.java:395)
        at junitparams.JUnitParamsRunner.runChild(JUnitParamsRunner.java:371)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:344)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:269)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:240)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:184)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:286)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:240)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
```
and here's the output from GraphicsMagick that suggests the image has CMYK color space:
```
Image: /opt/WP/Aplikacje/imageresizer/imageresizer-worker-client/target/test-classes/images/1309390204204_0x22111122_CMYK.jpeg
  Format: JPEG (Joint Photographic Experts Group JFIF format)
  Geometry: 591x522
  Class: DirectClass
  Type: color separated
  Depth: 8 bits-per-pixel component
  Channel Depths:
    Cyan:     8 bits
    Magenta:  8 bits
    Yellow:   8 bits
    Black:    8 bits
  Channel Statistics:
    Cyan:
      Minimum:                     0.00 (0.0000)
      Maximum:                 65535.00 (1.0000)
      Mean:                    28968.95 (0.4420)
      Standard Deviation:      25962.59 (0.3962)
    Magenta:
      Minimum:                 13878.00 (0.2118)
      Maximum:                 45746.00 (0.6980)
      Mean:                    28616.21 (0.4367)
      Standard Deviation:       6146.51 (0.0938)
    Yellow:
      Minimum:                 25700.00 (0.3922)
      Maximum:                 48316.00 (0.7373)
      Mean:                    36232.54 (0.5529)
      Standard Deviation:       5041.26 (0.0769)
    Black:
      Minimum:                     0.00 (0.0000)
      Maximum:                 65535.00 (1.0000)
      Mean:                    20478.94 (0.3125)
      Standard Deviation:      30337.49 (0.4629)
  Filesize: 92.2Ki
  Interlace: No
  Orientation: Unknown
  Background Color: white
  Border Color: #DFDFDF
  Matte Color: #BDBDBD
  Page geometry: 591x522+0+0
  Compose: Over
  Dispose: Undefined
  Iterations: 0
  Compression: JPEG
  JPEG-Quality: 80
  JPEG-Colorspace: 4
  JPEG-Colorspace-Name: CMYK
  JPEG-Sampling-factors: 2x2,1x1,1x1,2x2
  Signature: ff1ff41682c1e569e6b6c88566a2df629b212fda8b6f40f5c3569625fe5dcf24
  Tainted: False
  User Time: 0.010u
  Elapsed Time: 0m:0.006474s
  Pixels Per Second: 45.4Mi
```

**Comment by haraldk** [comment]
I think GraphicsMagick  is wrong about the color space. ;-)

Interesting stack trace though... 
Can you also tell me what Java version you are using (output of `java -version`)? And on which platform? 

-- 
Harald K

**Comment by pszemus** [comment]
I created a simple maven project just to check this. `pom.xml` has only one depencency:
```
		<dependency>
			<groupId>com.twelvemonkeys.imageio</groupId>
			<artifactId>imageio-jpeg</artifactId>
			<version>3.3.2</version>
		</dependency>
```
and the output is as folows:
```
reader: com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader@506c589e
reader: com.sun.imageio.plugins.jpeg.JPEGImageReader@446cdf90
Exception in thread "main" javax.imageio.IIOException: Unsupported Image Type
	at com.sun.imageio.plugins.jpeg.JPEGImageReader.readInternal(JPEGImageReader.java:1068)
	at com.sun.imageio.plugins.jpeg.JPEGImageReader.read(JPEGImageReader.java:1039)
	at com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader.read(JPEGImageReader.java:382)
	at javax.imageio.ImageIO.read(ImageIO.java:1448)
	at javax.imageio.ImageIO.read(ImageIO.java:1308)
	at pl.wp.imageiotest.ImageIOTest.main(ImageIOTest.java:17)
```

```
$ java -version
openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-b14)
OpenJDK 64-Bit Server VM (build 25.121-b14, mixed mode)
$ uname -a
Linux pszemus 4.10.8-100.fc24.x86_64 #1 SMP Fri Mar 31 13:20:57 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by haraldk** [comment]
Very strange indeed. I'm using the exact same Java version (1.8.0_121), but the official Oracle build on Windows. Works fine. I'll test on OS X when I get home.

There has been reports about build issues on OpenJDK before, but not runtime issues. But if you can try a different JRE build (Oracle), it's worth checking out.

-- 
Harald K

**Comment by pszemus** [comment]
You're right, no problem with JRE from Oracle:
```
$ /tmp/jre1.8.0_121/bin/java -version
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)
```
output:
```
reader: com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader@69d0a921
reader: com.sun.imageio.plugins.jpeg.JPEGImageReader@799f7e29
```

**Comment by haraldk** [comment]
Could be something about the libjpeg version in use. I think the Oracle JRE always uses its own bundled libjpeg version, but OpenJDK can use the system default libjpeg. I'm afraid you have to dig a little on your own, as I don't have a Linux machine (might try OpenJDK at home though).

I assume you can read other JPEGs using the OpenJDK build?

-- 
Harald K

**Comment by pszemus** [comment]
Thanks, I'll check which versions of libjpeg both JRE use.

> I assume you can read other JPEGs using the OpenJDK build?

Yes, only that particular one is being problematic.

**Comment by haraldk** [comment]
No problem on MacOS either.. 

One thing you could try though, if you don't mind building from source... In `JPEGImageReader.java` after line 381 in the `read` method, insert the following:

````java
                !delegate.getImageTypes(imageIndex).hasNext() ||
````

Then, unfortunately, I found a bug in the current code, so you need to patch some more, in the `readImageAsRasterAndReplaceColorProfile` method, around line 502, change:

````java
            // Apply source color conversion from implicit color space
            if (csType == JPEGColorSpace.YCbCr || csType == JPEGColorSpace.YCbCrA) {
                convertYCbCr2RGB(raster);
            }
            else if (csType == JPEGColorSpace.YCCK) {
````

to 

````java
            // Apply source color conversion from implicit color space
            if (csType == JPEGColorSpace.YCbCr) {
                convertYCbCr2RGB(raster, 3);
            }
            else if (csType == JPEGColorSpace.YCbCrA) {
                convertYCbCr2RGB(raster, 4);
            }
            else if (csType == JPEGColorSpace.YCCK) {
````

And finally replace the entire `convertYCbCr2RGB` method with:

````java
    public static void convertYCbCr2RGB(final Raster raster, final int numComponents) {
        final int height = raster.getHeight();
        final int width = raster.getWidth();
        final byte[] data = ((DataBufferByte) raster.getDataBuffer()).getData();

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                YCbCrConverter.convertYCbCr2RGB(data, data, (x + y * width) * numComponents);
            }
        }
    }
````

I haven't been able to test, obviously, but in theory it should work. ;-)

-- 
Harald K

**Comment by pszemus** [comment]
Hi
Thank you for your help. Unfortunately I wasn't able to build the TwelveMonkeys project without skipping tests (another OpenJDK case) but when I finally did, your changes fixed the problem and now the sample file is read correctly by both JREs.
Will those changes be merged with upstream and released anytime soon?

**Comment by haraldk** [comment]
Great!

Thanks for testing! I'll try to integrate this into the next release.

Sorry, should have warned you about the OpenJDK build, it's a known issue (#283). 

-- 
Harald K

**Comment by haraldk** [comment]
Fixed in 2eecdbb
